From e3c8171d3904c04a09b4bc5fc556dad131d84aeb Mon Sep 17 00:00:00 2001
From: frosty <gabriel@bwaaa.monster>
Date: Wed, 22 Apr 2026 00:04:03 -0400
Subject: [PATCH] fix: don't hardcode lua header path

---
 include/lua_build.hh | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/include/lua_build.hh b/include/lua_build.hh
index 1a7df06..a7cdf39 100644
--- a/include/lua_build.hh
+++ b/include/lua_build.hh
@@ -1,9 +1,9 @@
 #ifndef LUA_BUILD
 #define LUA_BUILD
 
-#include <lua5.1/lua.h>
-#include <lua5.1/lauxlib.h>
-#include <lua5.1/lualib.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
 
 bool lua_build(const char *path);
 
-- 
2.52.0

